File: Clients/Html5Q/HellouQ/qframework/ObjectsFactory.js
/*
   Copyright 2012, Telum Slavonski Brod, Croatia.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
   
   This file is part of QFramework project, and can be used only as part of project.
   Should be used for peace, not war :)   
*/

function ObjectsFactory(app)
{
	this.mApp = app;
	this.mItems = {};
}
	

ObjectsFactory.prototype.get = function(name)
{
	var model = this.mItems[name];
	return model;

}

ObjectsFactory.prototype.addModel = function( name, item)
{
	if ( this.mItems[name] != undefined)
	{
		return;
	}
	this.mApp.mWorld.add(item.mModel);
	item.mModel.mEnabled = true;
	this.mItems[name] =  item;
}

ObjectsFactory.prototype.removeModel = function( name)
{
	if (this.mItems[name] == undefined)
	{
		return;
	}	
	var item = this.mItems[name];

	var model = item.mModel;
	model.setVisible(false);

	delete this.mItems[name];
	this.mApp.mWorld.remove(model);
	// todo check if model hangs on in world
}

ObjectsFactory.prototype.create = function(name, data) 
{
	if ( this.mItems[name] != undefined)
	{
		return;
	}
	
	
	var model = this.mApp.mItems.getFromTemplate(name,data);
	if (model != undefined)
	{
		var item = new LayoutItem(this.mApp);
		item.mModel = model;
		this.addModel(name,item);
	}
}	

ObjectsFactory.prototype.place = function(name, data) 
{
	var item = this.mItems[name];
	if (item == undefined)
	{
		return;
	}
	var model = item.mModel;

	// TODO submodels
	if (model.mRefs.length == 0)
	{
		var ref = new GameonModelRef(model);
		model.addref(ref);
		item.mModelRef = ref;
	}
	// parse loc
	var tok = data.split(",");
	var x =0 ,y =0 ,z =0 ;
	
	if (data.length > 0)x = parseFloat(tok[0]);
	if (data.length > 1)y = parseFloat(tok[1]);
	if (data.length > 2)z = parseFloat(tok[2]);
	item.setPosition2(x,y,z);

}

ObjectsFactory.prototype.scale = function(name, data) 
{
	var item = this.mItems[name];
	if (item == undefined)
	{
		return;
	}
	var model = item.mModel;

	
	// TODO submodels
	if (model.mRefs.length == 0)
	{
		var ref = new GameonModelRef(model);
		model.addref(ref);
	}
	// parse loc
	var tok = data.split(",");
	var x =0 ,y =0 ,z =0 ;
	
	if (data.length > 0)x = parseFloat(tok[0]);
	if (data.length > 1)y = parseFloat(tok[1]);
	if (data.length > 2)z = parseFloat(tok[2]);
	var r = model.ref(0);
	r.setScale(x, y, z);
	r.set();
}

ObjectsFactory.prototype.texture = function(name, data) 
{
	var item = this.mItems[name];
	if (item == undefined)
	{
		return;
	}
	
	var model = item.mModel;
	
	var text = this.mApp.mTextures.getTexture(data);
	model.setTexture(text);
}
//TODO mutliple references with name.refid , default 0!

ObjectsFactory.prototype.state = function(name, data) 
{
	var item = this.mItems[name];
	if (item == undefined)
	{
		return;
	}
	
	var model = item.mModel;

	var visible = false;
	if (data == "visible")
	{
		visible = true;
	}
	if (model.mRefs.length == 0)
	{
		this.place(name, "0,0,0");
	}
	
	model.ref(0).setVisible(visible);
	model.setVisible(visible);
}


ObjectsFactory.prototype.remove = function(name, data) 
{
	var item = this.mItems[name];
	if (item == undefined)
	{
		return;
	}
	var model = item.mModel;
	this.removeModel(name);
}		


ObjectsFactory.prototype.initObjects = function(response)
{
	var areas;
	areas = response["object"];
	for (var a=0; a< areas.length; a++)
	{
		//ServerkoAreaData* pCurr = [response.mAreaDatas objectAtIndex:a];
		var pCurr = areas[a];
		this.processObject(pCurr);
	}	

}

ObjectsFactory.prototype.processObject = function( objData) {
	var name = objData["name"];
	var template = objData["template"];
	this.create(name , template);
	
	if (objData["location"] != undefined)
	{	        
		var data = objData["location"];    
		this.place(name, data);
	}
	
	if (objData["bounds"] != undefined)
	{	        
		var data = objData["bounds"];    
		this.scale(name, data);
	}
	
	if (objData["texture"] != undefined)
	{	        
		var data = objData["texture"];    
		this.texture(name, data);
	}
	
	if (objData["state"] != undefined)
	{	        
		var data = objData["state"];    
		this.state(name, data);
	}	

}

File: Clients/JavaQGl/Q/src/com/qframework/core/ObjectsFactory.java
/*
   Copyright 2012, Telum Slavonski Brod, Croatia.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
   
   This file is part of QFramework project, and can be used only as part of project.
   Should be used for peace, not war :)   
*/

package com.qframework.core;

import java.util.HashMap;
import java.util.NoSuchElementException;
import java.util.StringTokenizer;

import javax.media.opengl.GL2;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class ObjectsFactory {

	private HashMap<String, LayoutItem>		mItems = new HashMap<String, LayoutItem>();
	private GameonApp 	mApp;
	
	ObjectsFactory(GameonApp parent)
	{
		mApp = parent;
	}

	public LayoutItem get(String name)
	{
		LayoutItem model = mItems.get(name);
		return model;
	}
	void addModel( String name, LayoutItem item)
	{
		if ( mItems.containsKey(name))
		{
			return;
		}
		item.mModel.mEnabled = true;
		mApp.world().add(item.mModel);
		mItems.put(name, item);
	}
	
	void removeModel( String name)
	{
		if (! mItems.containsKey(name))
		{
			return;
		}	
		LayoutItem item = mItems.get(name);
	
		GameonModel model = item.mModel;
		model.setVisible(false);
		mItems.remove( name);
		mApp.world().remove(model);
		// todo check if model hangs on in world
	}

	public void create(String name, String data) {
		if ( mItems.containsKey(name))
		{
			return;
		}
		
		
		GameonModel model = mApp.items().getFromTemplate(name,data);
		if (model != null)
		{
			LayoutItem item = new LayoutItem(mApp);
			item.mModel = model;
			addModel(name,item);
		}
	}	
	public void place(String name, String data) {
		LayoutItem item = mItems.get(name);
		if (item == null)
		{
			return;
		}
		GameonModel model = item.mModel;

		// TODO submodels
		if (model.ref(0) == null)
		{
			GameonModelRef ref = new GameonModelRef(model);
			model.addref(ref);
			item.mModelRef = ref;
		}
		// parse loc
        StringTokenizer tok = new StringTokenizer(data, ",");
        try {
        	float x =0 ,y =0 ,z =0 ;
            try {
                x = Float.parseFloat(tok.nextToken());
            } catch (NumberFormatException e) {
            }
            try {
                y = Float.parseFloat(tok.nextToken());
            } catch (NumberFormatException e) {
            }
            try {
                z = Float.parseFloat(tok.nextToken());
            } catch (NumberFormatException e) {
            }
            item.setPosition2(x,y,z);
        } catch (NoSuchElementException e) {
        }
		

	}
	public void scale(String name, String data) {
		LayoutItem item = mItems.get(name);
		if (item == null)
		{
			return;
		}
		GameonModel model = item.mModel;
	
		
		// TODO submodels
		if (model.ref(0) == null)
		{
			GameonModelRef ref = new GameonModelRef(model);
			model.addref(ref);
		}
		// parse loc
        StringTokenizer tok = new StringTokenizer(data, ",");
        try {
        	float x =0 ,y =0 ,z =0 ;
            try {
                x = Float.parseFloat(tok.nextToken());
            } catch (NumberFormatException e) {
            }
            try {
                y = Float.parseFloat(tok.nextToken());
            } catch (NumberFormatException e) {
            }
            try {
                z = Float.parseFloat(tok.nextToken());
            } catch (NumberFormatException e) {
            }
            GameonModelRef r = model.ref(0);
            r.setScale(x, y, z);
            r.set();
        } catch (NoSuchElementException e) {
        }		
	}
	
	public void rotate(String name, String data) {
		LayoutItem item = mItems.get(name);
		if (item == null)
		{
			return;
		}
		GameonModel model = item.mModel;
	
		
		// TODO submodels
		if (model.ref(0) == null)
		{
			GameonModelRef ref = new GameonModelRef(model);
			model.addref(ref);
		}
		
		float[] vals = new float[3];
		ServerkoParse.parseFloatArray(vals , data );
        GameonModelRef r = model.ref(0);
        r.setRotate(vals);
        r.set();
	}

	
	public void texture(String name, String data) {
		LayoutItem item = mItems.get(name);
		if (item == null)
		{
			return;
		}
		GameonModel model = item.mModel;
		
		int text = mApp.textures().getTexture(data);
		model.setTexture(text);
	}
	
	//TODO mutliple references with name.refid , default 0!
	public void state(String name, String data) {
		LayoutItem item = mItems.get(name);
		if (item == null)
		{
			return;
		}
		GameonModel model = item.mModel;

		boolean visible = false;
		if (data.equals("visible"))
		{
			visible = true;
		}
		if (model.ref(0) == null)
		{
			this.place(name, "0,0,0");
		}
		
		model.ref(0).setVisible(visible);
		model.setVisible(visible);
	}
	
	public void remove(String name, String data) {
		LayoutItem item = mItems.get(name);
		if (item == null)
		{
			return;
		}
		GameonModel model = item.mModel;
		removeModel(name);
	}
	
	public void initObjects(GL2 gl, JSONObject response)
    {
        // init layout
		try {
	    	JSONArray areas;
			areas = response.getJSONArray("object");
        
	        for (int a=0; a< areas.length(); a++)
	        {
	            JSONObject pCurr = areas.getJSONObject(a);
	            processObject(gl, pCurr);
	        }
		} catch (JSONException e) {
			e.printStackTrace();
        }

	}

    private void processObject(GL2 gl, JSONObject objData) {
    	try {
			String name = objData.getString("name");
			String template = objData.getString("template");
			
			create(name , template);
			
			if (objData.has("location"))
			{
				String data = objData.getString("location");
				place(name, data);
			}
			if (objData.has("bounds"))
			{
				String data = objData.getString("bounds");
				scale(name, data);
			}			
			if (objData.has("texture"))
			{
				String data = objData.getString("texture");
				texture(name, data);
			}			
			if (objData.has("state"))
			{
				String data = objData.getString("state");
				state(name, data);
			}			
    	}
    	catch (JSONException e) {
    		e.printStackTrace();
        }    	
    }
    
}
